Constructor to global variable in converter generates illegal c code

This is a tad difficult to explain so I am going to give examples. This generates illegal c code:

``` nim
type
  Inner =  object
    foo: string
converter toString(i: int): string =
  var g{.global.} = Inner(foo: $i)
var s: string = 4
```

> src\nimcache\converter_test.c: In function 'converter_testInit000':
> src\nimcache\converter_test.c:354:30: error: 'i0' undeclared (first use in this function)  
> LOC100001.foo = nimIntToStr(i0);
>                               ^
> src\nimcache\converter_test.c:354:30: note: each undeclared identifier is reported only once for each  function it appears in

In contrast this does compile and work as expected:

``` nim
type
  Inner =  object
    foo: string
converter toString(i: int): string =
  var g{.global.} : Inner
   g.foo = $i
var s: string = 4
```

When the string is constant, say Inner(foo: "bar"), or the converter is invoked explicitely the program also works.

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -727,9 +727,6 @@ template isLocalSym(sym: PSym): bool =
       sym.kind in {skProc, skFunc, skIterator} and
       sfGlobal notin sym.flags
 
-template isLocalVarSym(n: PNode): bool =
-  n.kind == nkSym and isLocalSym(n.sym)
-
 proc usesLocalVar(n: PNode): bool =
   case n.kind
   of nkSym:
@@ -754,7 +751,7 @@ proc usesLocalVar(n: PNode): bool =
     result = false
 
 proc globalVarInitCheck(c: PContext, n: PNode) =
-  if n.isLocalVarSym or usesLocalVar(n):
+  if usesLocalVar(n):
     localError(c.config, n.info, errCannotAssignToGlobal)
 
 const

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -731,16 +731,30 @@ template isLocalVarSym(n: PNode): bool =
   n.kind == nkSym and isLocalSym(n.sym)
 
 proc usesLocalVar(n: PNode): bool =
-  result = false
-  for z in 1 ..< n.len:
-    if n[z].isLocalVarSym:
-      return true
-    elif n[z].kind in nkCallKinds:
-      if usesLocalVar(n[z]):
+  case n.kind
+  of nkSym:
+    result = isLocalSym(n.sym)
+  of nkCallKinds, nkObjConstr:
+    result = false
+    for i in 1 ..< n.len:
+      if usesLocalVar(n[i]):
         return true
+  of nkTupleConstr, nkPar, nkBracket, nkCurly:
+    result = false
+    for i in 0 ..< n.len:
+      if usesLocalVar(n[i]):
+        return true
+  of nkDotExpr, nkCheckedFieldExpr,
+       nkBracketExpr, nkAddr, nkHiddenAddr,
+       nkObjDownConv, nkObjUpConv:
+    result = usesLocalVar(n[0])
+  of nkHiddenStdConv, nkHiddenSubConv, nkCast, nkExprColonExpr:
+    result = usesLocalVar(n[1])
+  else:
+    result = false
 
 proc globalVarInitCheck(c: PContext, n: PNode) =
-  if n.isLocalVarSym or n.kind in nkCallKinds and usesLocalVar(n):
+  if n.isLocalVarSym or usesLocalVar(n):
     localError(c.config, n.info, errCannotAssignToGlobal)
 
 const

